bestie/server/bigquery_metrics.go

```diff
@@ -0,0 +1,288 @@
+		}
+	} else {
+		fmt.Fprintf(w, "Created table %s\n", t.formatTableId())
+	}
+	return nil
+}
+
+// Translate protobuf metric to BigQuery metric.
+func translateMetric(stream bazelStream, m testMetric) (bigQueryMetric, error) {
+	// Process the slice of key/value pairs representing individual metric tags.
+	dat := make(map[string]string)
+	for _, kv := range m.tags {
+		dat[kv.key] = kv.value
+	}
+	// Insert additional tags using information that is only available
+	// to the BES endpoint through the Bazel event message itself.
+	//
+	// Storing both the invocationId and the invocationSha in the database,
+	// the former being useful to match up with build log references, etc.
+	dat["invocation_id"] = stream.invocationId
+	dat["invocation_sha"] = stream.invocationSha
+	dat["run"] = stream.run
+	dat["test_name"] = stream.testName
+	tags, err := json.Marshal(dat)
+	if err != nil {
+		return bigQueryMetric{}, fmt.Errorf("Error converting JSON to string: %s", err)
+	}
+
+	// Change the metric timestamp from integer to formatted string,
+	// since that is what BigQuery expects for a TIMESTAMP column.
+	//
+	// The incoming timestamp is UTC nanoseconds since epoch, but
+	// storing with microsecond granularity in BigQuery.
+	ts := m.timestamp / 1000
+	tsf := time.Unix(ts/1000000, ts%1000000).Format("2006-01-02 15:04:05.000000")
+
+	return bigQueryMetric{
+		metricname: m.metricname,
+		tags:       string(tags),
+		value:      m.value,
+		timestamp:  tsf,
+	}, nil
+}
+
+// Upload this set of metrics to the specified BigQuery table.
+func uploadTestMetrics(w io.Writer, stream bazelStream, r *metricTestResult) error {
+	// Normalize the BigQuery table identifier based on whether one was specified
+	// in the protobuf message.
+	normalizeTableRef(&r.table)
+	fmt.Fprintf(w, "Normalized table ref: %s\n", r.table.formatTableId())
+
+	// Get client context for this BigQuery operation.
+	ctx := context.Background()
+	client, err := bq.NewClient(ctx, r.table.project)
+	if err != nil {
+		return fmt.Errorf("bigquery.NewClient: %v", err)
+	}
+	defer client.Close()
+	session := bigQuerySession{ctx: ctx, client: client}
+
+	// Check if the BigQuery dataset and table exists; create each as needed.
+	if exist := r.table.isTableExist(session); !exist {
+		ServiceStats.incrementBigQueryTableNotFound()
+		if err := createTable(w, r.table, session); err != nil {
+			return err
+		}
+	}
+
+	// Prepare the metric rows for uploading to BigQuery.
+	// Make sure each row element references its own array item.
+	// Note: bigQueryMetric implements the ValueSaver interface.
+	var bqMetrics []bigQueryMetric
+	var rows []*bigQueryMetric
+	idx := 0
+	for _, m := range r.metrics {
+		// TODO: Decide whether to override client metric time with BES server current time.
+		bqMetric, err := translateMetric(stream, m)
+		if err != nil {
+			ServiceStats.incrementBigQueryMetricDiscard(1)
+			fmt.Fprintf(w, "Discarding metric %s due to error: %s\n", m, err)
```

Comment:
Addressing this in the following ways:
1. Created a logger to replace the fmt.Printf's, etc.
2. Defined debugPrintf() and debugPrintln() functions that call the logger if the isDebugMode global variable is set to true, which comes from a new `--debug` command line arg. Decided which are normal vs. debug log messages.
3. Used strings.Builder to create a single string to be logged / debug logged all at once, rather than logging it piecemeal.